src/slurm_plugin/common.py

```diff
@@ -8,5 +8,183 @@
 # or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
 # OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
 # limitations under the License.
+import collections
+import logging
+import re
+import subprocess
+
+import boto3
+from botocore.exceptions import ClientError
+
+from common.schedulers.slurm_commands import update_nodes
+from common.utils import grouper
 
 CONFIG_FILE_DIR = "/opt/parallelcluster/configs/slurm"
+
+log = logging.getLogger(__name__)
+
+
+def terminate_associated_instances(slurm_nodes, region, cluster_name, boto3_config, batch_size):
+    """Terminate instances associated with given nodes in batches."""
+    instance_ids_to_nodename = _get_instance_ids_to_nodename(slurm_nodes, region, cluster_name, boto3_config)
+    log.debug("instance_ids_to_nodename = %s", instance_ids_to_nodename)
+    if instance_ids_to_nodename:
+        delete_instances(
+            list(instance_ids_to_nodename.keys()), region, boto3_config, batch_size,
+        )
+
+
+def _get_instance_ids_to_nodename(slurm_nodes, region, cluster_name, boto3_config):
+    """Retrieve dict that maps from instance ids to slurm nodenames."""
+    node_ip_to_name = {node.nodeaddr: node.name for node in slurm_nodes}
+    ec2_client = boto3.client("ec2", region_name=region, config=boto3_config)
+    paginator = ec2_client.get_paginator("describe_instances")
+    response_iterator = paginator.paginate(
+        Filters=[
+            {"Name": "private-ip-address", "Values": list(node_ip_to_name.keys())},
+            {"Name": "tag:ClusterName", "Values": [cluster_name]},
+        ],
+    )
+    filtered_iterator = response_iterator.search("Reservations[].Instances[]")
+    return {
+        instance_info["InstanceId"]: node_ip_to_name[instance_info["PrivateIpAddress"]]
+        for instance_info in filtered_iterator
+    }
+
+
+def delete_instances(instance_ids_to_terminate, region, boto3_config, batch_size):
+    """Terminate corresponding EC2 instances."""
+    ec2_client = boto3.client("ec2", region_name=region, config=boto3_config)
+    log.info("Terminating instances %s", instance_ids_to_terminate)
+    for instances in grouper(instance_ids_to_terminate, batch_size):
+        try:
+            # Boto3 clients retries on connection errors only
+            ec2_client.terminate_instances(InstanceIds=list(instances),)
+        except ClientError as e:
+            log.error("Failed when terminating instances %s with error %s", instances, e)
+
+
+class InstanceLauncher:
```

Comment:
What I don't like about this class is that you can use it only once, meaning you cannot initialize an InstanceLauncher object and then call launch_instances (or add_instances_for_nodes) multiple times. Can we slightly change it so that we keep at the class level the data that would not change across multiple invocations (e.g. region, cluster_name..) and move attributes such as node_list to the actual method that gets called to launch new instances. Also failed_nodes needs to be re-initialized.

Also would it make sense to make this more generic and include the termination functionalities defined above? Also the module name could be updated since here we are grouping all functions related to instance management if I'm not wrong

Reply:
Regarding InstanceLauncher is not about performance but about limiting a class to be used only once without even enforcing this. I understand the reason why this class was created in the first place, which is mostly to have a shared state across the different methods, but I don't think this is something that prevents us from having the class behavior consistent across multiple invocations. If we don't want to do that, for the correctness of the class API we should at least raise an exception when the method to launch instance is invoked the second time. Personally I'd just reset the shared state when the method is invoked and also we could consider returning the list of failed nodes rather than reading it separately from the class attributes, making it an internal state only.

For the termination it was more for consistency but I'm fine leaving it as is.

Reply:
Ok that makes sense, I will revise InstanceLauncher, and see if I can potentially include termination functions as well. If not I will just leave termination functions as is.